runtime.md

```diff
@@ -23,37 +14,127 @@ This allows the hooks to perform cleanup and teardown logic after the runtime de
+Unless otherwise stated, generating an error MUST leave the state of the environment as if the operation were never attempted - modulo any possible trivial ancillary changes such as logging.
+
+### Query State
+
+`state <container-id>`
+
+This operation MUST generate an error if it is not provided the ID of a container.
+This operation MUST return the state of a container as specified in the [State](#state) section.
+
+
+### Start
+
+`start <path-to-bundle> <container-id>`
+
+This operation MUST generate an error if it is not provided a path to the bundle and the container ID to associate with the container.
+If the ID provided is not unique across all containers running on this host, or is not valid in any other way, the implementation MUST generate an error.
+Using the data in `config.json` and `runtime.json` files, that are in the bundle's directory, this operation MUST create a new container.
+This includes creating the relevant namespaces, resource limits, etc and configuring the appropriate capabilities for the container.
+A new process within the scope of the container  MUST be created as specified by the `config.json` file.
+The current working directory of the process MUST be set to the `cwd` field of the config.
+If the path specified by the `cwd` fields does not exist then an error MUST be generated.
+
+Attempting to start an already running container MUST have no effect on the container and MUST generate an error.
+
+### Stop
+
+`stop <container-id>`
+
+This operation MUST generate an error if it is not provided the container ID.
+This operation MUST stop and delete a running container.
+Stopping a container MUST stop all of the processes running within the scope of the container.
+Deleting a container MUST delete the associated namespaces and resources associated with the container.
+Once a container is deleted, its `id` MAY be used by subsequent containers.
+Attempting to stop a container that is not running MUST have no effect on the container and MUST generate an error.
+
+### Pause
+
+`pause <container-id>`
+
+This operation MUST generate an error if it is not provided the container ID.
+This operation MUST suspend all processes that are running within the container.
+If the container is not running, either stopped or aleady paused, then this operation MUST have no effect on the container and MUST generate an error.
+
+### Unpause
+
+`unpause <container-id>`
+
+This operation MUST generate an error if it is not provided the container ID.
+This operation MUST resume all processes that were previously paused via the `pause` operation.
+If the container is not paused then this operation MUST have no effect on the container and MUST generate an error.
+
+### Exec
+
+`exec <container-id> <path-to-json>`
+
+This operation MUST generate an error if it is not provided the container ID and a path to the JSON describing the process to start.
```

Comment:
On Thu, Jan 21, 2016 at 10:36:34AM -0800, Doug Davis wrote:
> > +This operation MUST generate an error if it is not provided the
> > container ID and a path to the JSON describing the process to
> > start.
>
> doesn't the test below cover it?

It's enough to work from, especially since it looks a lot like [1].
But I'd rather explicitly reference [1] and say “the config JSON MUST
contain a ‘process’ entry as it's root object” and then soften your
“MUST adhere to the following format” to an informative example.

See also [2], where I argue for a single process schema that we use
wherever we need process information (vs. a custom process schema for
each location).

[1]: https://github.com/opencontainers/specs/blob/v0.2.0/config.md#process-configuration
[2]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/Cn5L6g1prgA
     Subject: Unify container-process and hook-process structures
     Date: Mon, 4 Jan 2016 21:56:37 -0800
     Message-ID: <20160105055637.GG6362@odin.tremily.us>